src/core/SampleLoader.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+ * SampleLoader.cpp - Sample loader with support for caching
+ *
+ * Copyright (c) 2023 saker <sakertooth@gmail.com>
+ *               2024 Dalton Messmer <messmer.dalton/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleLoader.h"
+
+#include <cassert>
+
+#include "Engine.h"
+#include "Song.h"
+#include "lmms_basics.h"
+
+namespace lmms {
+
+namespace {
+
+void displayError(const QString& message)
+{
+	Engine::getSong()->collectError(QString("%1: %2").arg(QObject::tr("Error loading sample"), message));
+}
+
+} // namespace
+
+/**
+ * An evictor which automatically evicts sample buffers from the cache
+ * immediately after the last reference to the buffer is dropped.
+ * Used as std::shared_ptr<const SampleBuffer>'s deleter.
+*/
+struct SampleLoader::AutoEvictor
+{
+	void operator()(SampleBuffer* p) const noexcept
+	{
+		if (!p) { return; }
+
+		instance().remove(*p);
```

Comment:
It has been a while since I touched the lmms code, but I think it is a big dangerous for thread safety to call remove() directly here (and not via signal/slot). Since we have no idea (and no guarantee) about which thread would be the last thread to use the sample buffer.

Another point I am unsure about is what would happen if another singleton would hold a cached sample buffer while closing the program. I am not sure about the rules of using singletons from inside of another singleton's destructor.

Reply:
Cache accesses should be thread safe as of [e9922d4](https://github.com/LMMS/lmms/pull/7058/commits/e9922d4f0b8896083ba0166b4ed97b8048f17761).